source/extensions/http/stateful_session/cookie/cookie.h

```diff
@@ -52,20 +62,55 @@ class CookieBasedSessionStateFactory : public Envoy::Http::SessionStateFactory {
   }
 
 private:
-  absl::optional<std::string> parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+  std::pair<absl::optional<std::string>, bool>
+  parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+    bool use_old_style_encoding = false;
     const std::string cookie_value = Envoy::Http::Utility::parseCookieValue(headers, name_);
-    std::string address = Envoy::Base64::decode(cookie_value);
+    if (cookie_value.empty()) {
+      return std::make_pair(absl::nullopt, use_old_style_encoding);
+    }
+    const std::string decoded_value = Envoy::Base64::decode(cookie_value);
+    std::string address;
+
+    // Try to interpret the cookie as proto payload.
+    // Otherwise treat it as "old" style format, which is ip-address:port.
+    envoy::Cookie cookie;
+    if (cookie.ParseFromString(decoded_value)) {
+      address = cookie.address();
```

Comment:
Would it be better to add a prefix flag to tell the cookie value is new format more explicitly? For example the `PB <base64 encoded proto>` is new format and `base64 encoded address` is old format.

Reply:
I am not sure if it will bring much value. The problem is that interpreting cookie opens another possibility for an attack. Even if the cookie had `PB <base64 encoded proto>` format, the code cannot blindly assume that the content is the correctly encoded proto. It will still have to validate the wire format and reject it when de-serialization fails. So, here if it is not proto, try to interpret it as "old" format. When in the future we decide to deprecate "old" format, the code will ignore the cookie all together instead of trying to use "old" format.

Reply:
I think this PB prefix just a simple flag to tell envoy to run which logic explicitly. We needn't try to prasing the value as proto first at every time. And at least it will not make things worse.

Reply:
I truly think it is unnecessary, especially if the "old" format is going to be deprecated. Unless someone hacks the cookie, the is no way that Envoy will ever receive "old" cookie, because it will never generate it. I also have to check rfc defining cookies. I am not sure if spaces are allowed.

Reply:
> Unless someone hacks the cookie, the is no way that Envoy will ever receive "old" cookie, because it will never generate it

I think these may be existing cookies will be sent, at the timepoint of upgrade envoy to new version?

And I still think a simple plain format cookie support would be helpful for the test.

Reply:
>  I also have to check rfc defining cookies. I am not sure if spaces are allowed.

🙏 Thanks for all your patience and these hard working. Hope we can get a thorough design as far as possible.